File: java/src/main/java/neetcode/slidingwindow/p0567_permutationinstring/Solution.java
package neetcode.slidingwindow.p0567_permutationinstring;

import java.util.Arrays;

public class Solution {

    /**
     * Returns true if any permutation of s1 exists as a contiguous
     * substring of s2.
     *
     * Approach: fixed-size sliding window of length s1.length().
     * A window is a permutation of s1 iff their character frequency
     * arrays are identical. Maintain a frequency array for the current
     * window: add the incoming character on the right, remove the
     * outgoing character on the left, compare with s1's array.
     *
     * Time:  O(n) where n = s2.length() — each character is visited once;
     *        the array comparison is O(26) = O(1)
     * Space: O(1) — two fixed 26-entry arrays
     */
    public boolean checkInclusion(String s1, String s2) {
        if (s1.length() > s2.length()) return false;

        int[] countS1 = new int[26];
        int[] countWin = new int[26];
        int len = s1.length();

        // Initialize both frequency arrays with the first window
        for (int i = 0; i < len; i++) {
            countS1[s1.charAt(i) - 'a']++;
            countWin[s2.charAt(i) - 'a']++;
        }

        // Check the first window
        if (Arrays.equals(countS1, countWin)) return true;

        // Slide the window one step at a time
        for (int right = len; right < s2.length(); right++) {
            countWin[s2.charAt(right) - 'a']++;           // new char enters
            countWin[s2.charAt(right - len) - 'a']--;     // old char leaves
            if (Arrays.equals(countS1, countWin)) return true;
        }

        return false;
    }
}

File: python/src/neetcode/slidingwindow/p0567_permutationinstring/solution.py
from collections import Counter
from typing import List

class Solution:
    """
    Fixed-size sliding window using Counter comparison.
    Removes keys with count 0 to keep Counter comparison correct.

    Time:  O(n) — each character enters and leaves the window once
    Space: O(1) — at most 26 keys in each Counter
    """
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if len(s1) > len(s2):
            return False

        count_s1 = Counter(s1)
        count_win = Counter(s2[:len(s1)])

        if count_s1 == count_win:
            return True

        for right in range(len(s1), len(s2)):
            count_win[s2[right]] += 1

            left_char = s2[right - len(s1)]
            count_win[left_char] -= 1
            if count_win[left_char] == 0:
                del count_win[left_char]      # avoid {char: 0} breaking equality

            if count_win == count_s1:
                return True

        return False


class SolutionArray:
    """
    Fixed-size sliding window using int[26] arrays — mirrors the Java solution.
    No key-cleanup needed: list == list compares element by element.

    Time:  O(n)
    Space: O(1) — two fixed arrays of 26 integers
    """
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if len(s1) > len(s2):
            return False

        count_s1 = [0] * 26
        count_win = [0] * 26

        for c in s1:
            count_s1[ord(c) - ord('a')] += 1
        for c in s2[:len(s1)]:
            count_win[ord(c) - ord('a')] += 1

        if count_s1 == count_win:
            return True

        for right in range(len(s1), len(s2)):
            count_win[ord(s2[right]) - ord('a')] += 1
            count_win[ord(s2[right - len(s1)]) - ord('a')] -= 1
            if count_s1 == count_win:
                return True

        return False
